src/test/java/org/elasticsearch/search/SearchWithRejectionsTests.java

```diff
@@ -0,0 +1,96 @@
+ * under the License.
+ */
+
+package org.elasticsearch.search;
+
+import com.google.common.base.Predicate;
+import org.apache.lucene.util.LuceneTestCase;
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsResponse;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.search.SearchType;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.test.ElasticsearchIntegrationTest;
+import org.junit.Test;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.common.settings.Settings.settingsBuilder;
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.SUITE)
+@LuceneTestCase.Slow
+public class SearchWithRejectionsTests extends ElasticsearchIntegrationTest {
+    @Override
+    public Settings nodeSettings(int nodeOrdinal) {
+        return settingsBuilder().put(super.nodeSettings(nodeOrdinal))
+                .put("threadpool.search.type", "fixed")
+                .put("threadpool.search.size", 1)
+                .put("threadpool.search.queue_size", 1)
+                .build();
+    }
+
+    @Test
+    public void testOpenContextsAfterRejections() throws InterruptedException {
+        createIndex("test");
+        ensureGreen("test");
+        final int docs = scaledRandomIntBetween(20, 50);
+        for (int i = 0; i < docs; i++) {
+            client().prepareIndex("test", "type", Integer.toString(i)).setSource("field", "value").execute().actionGet();
+        }
+        IndicesStatsResponse indicesStats = client().admin().indices().prepareStats().execute().actionGet();
+        assertThat(indicesStats.getTotal().getSearch().getOpenContexts(), equalTo(0l));
+        refresh();
+
+        int numSearches = 10;
+        Future<SearchResponse>[] responses = new Future[numSearches];
+        SearchType searchType = randomFrom(SearchType.DEFAULT, SearchType.QUERY_AND_FETCH, SearchType.QUERY_THEN_FETCH, SearchType.DFS_QUERY_AND_FETCH, SearchType.DFS_QUERY_THEN_FETCH);
+        logger.info("search type is {}", searchType);
+        for (int i = 0; i < numSearches; i++) {
+            responses[i] = client().prepareSearch()
+                    .setQuery(matchAllQuery())
+                    .setSearchType(searchType)
+                    .execute();
+        }
+        int failures = 0;
+        for (int i = 0; i < numSearches; i++) {
+            try {
+                SearchResponse searchResponse = responses[i].get();
+                if (searchResponse.getFailedShards() > 0) {
+                    failures++;
+                }
+            } catch (Throwable t) {
+                failures++;
+            }
+        }
+        assertThat(failures, greaterThanOrEqualTo(1));
```

Comment:
is this guaranteed? I mean if the cpu is fast/slow enough we might not reject anything in theory? I think we can omit this assertion?